### PR TITLE
8294047: HttpResponseInputStream swallows interrupts

### PR DESCRIPTION
**Pull request by DarraghClarke**
Currently if a `HttpResonseInputStream` gets interrupted while reading it will just swallow the exception and continue,

This PR changes it to close the stream and throw an IOException, I added a test to cover this which just uses two threads to read the stream then interrupt it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8294047](https://bugs.openjdk.org/browse/JDK-8294047): HttpResponseInputStream swallows interrupts
 * [JDK-8298163](https://bugs.openjdk.org/browse/JDK-8298163): HttpResponseInputStream swallows interrupts (**CSR**)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Vyom Tewari](https://openjdk.org/census#vtewari) (@vyommani - Committer) ⚠️ Review applies to [5fc55a3a](https://git.openjdk.org/jdk/pull/11323/files/5fc55a3ada831cdd396bee234971049f93369243)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11323/head:pull/11323` \
`$ git checkout pull/11323`

Update a local copy of the PR: \
`$ git checkout pull/11323` \
`$ git pull https://git.openjdk.org/jdk pull/11323/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11323`

View PR using the GUI difftool: \
`$ git pr show -t 11323`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11323.diff">https://git.openjdk.org/jdk/pull/11323.diff</a>

</details>

### CODE REVIEW

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 1]
```diff
@@ -481,7 +481,8 @@ private ByteBuffer current() throws IOException {
                     if (debug.on()) debug.log("Next Buffer");
                     currentBuffer = currentListItr.next();
                 } catch (InterruptedException ex) {
-                    // continue
+                    close();
+                    throw new IOException("interrupted", ex);
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:485]: I'd suggest to throw `InterruptedIOException` here instead.

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    @Test
+    public void test() {
+        try {
+            //create client and interrupter threads
+            Thread clientThread = createClientThread(countDownLatch, port);
+            Thread interrupterThread = new Thread(() -> {
+                try {
+                    countDownLatch.await();
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:87]: It would be useful to add a comment before waiting on the latch to explain what you are waiting for.

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            interrupterThread.start();
+            clientThread.join();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static class Handler implements HttpHandler {
+
+        CountDownLatch countDownLatch;
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:107]: Maybe a better name than `countDownLatch` could be found - that could give a clue on what this latch is used to wait for?

Comment (DarraghClarke) [reply test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:107]: I was thinking maybe `messageLatch`, `messageReceivedLatch` or `clientReadyLatch`

Comment (dfuch) [reply test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:107]: Maybe we rename both:
```
countDownLatch  => interruptReadyLatch
interruptLatch  => interruptDoneLatch
```

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    static final String SECOND_MESSAGE = "Shouldn't be received";
+
+    @BeforeAll
+    void before() throws Exception {
+        InetAddress loopback = InetAddress.getLoopbackAddress();
+        InetSocketAddress addr = new InetSocketAddress(loopback, 0);
+        server = HttpServer.create(addr, 0);
+        port = server.getAddress().getPort();
+        Handler handler = new Handler(countDownLatch, interruptLatch);
+        server.createContext("/", handler);
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:71]: It would be good to use a path unique for this test to avoid potential interferences. I would suggest registering the context under "/HttpResponseInputStreamInterruptTest/" instead of "/".

And use that when building the URI below as well.

This is important because the handler is using latches, and therefore can be invoked only once...

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 5]
```diff
@@ -481,8 +482,11 @@ private ByteBuffer current() throws IOException {
                     if (debug.on()) debug.log("Next Buffer");
                     currentBuffer = currentListItr.next();
                 } catch (InterruptedException ex) {
-                    close();
-                    throw new IOException("interrupted", ex);
+                    try {
+                        close();
+                    } catch (IOException ignored) {
+                    }
+                    throw new InterruptedIOException();
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:489]: Thanks for doing that. Maybe we should keep the caught InterruptedException as the cause of the new InterruptedIOException. I'd suggest to add a new utility method to the `...common.Utils` class for that. I mean - something like:
```
    InterruptedIOException Utils.toInterruptedIOException(InterruptedException ex);
```

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 6]
```diff
@@ -104,12 +107,12 @@ public void test() {
 
     static class Handler implements HttpHandler {
 
-        CountDownLatch countDownLatch;
-        CountDownLatch interruptLatch;
+        CountDownLatch interruptReadyLatch;
+        CountDownLatch interruptDoneLatch;
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:111]: should be `final`

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 7]
```diff
@@ -160,12 +163,11 @@ static Thread createClientThread(CountDownLatch countDownLatch, int port) {
                 assertEquals(firstOutput, FIRST_MESSAGE);
 
                 // countdown on latch, and assert that an IOException is throw due to the interrupt
-                countDownLatch.countDown();
-                assertThrows(IOException.class, () -> response.body().readAllBytes(), "excepted IOException");
+                interruptReadyLatch.countDown();
+                assertThrows(InterruptedIOException.class, () -> response.body().readAllBytes(), "excepted IOException");
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:167]: Maybe change the message while you're at it since we're actually expecting InterruptedIOException

**Review by dfuch — CHANGES_REQUESTED** [review]

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 5 above)
Comment (DarraghClarke) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:489]: So  something in utils like `return new InterruptedIOException(String.valueOf(exxeption));` ?

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:489]: No something that would call `initCause(ex)` to set the root cause - as AFAICS there's no constructor that takes a root cause.

Comment (DarraghClarke) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:489]: ah ok, so create a new `InterruptedIOException` and set the initcause to be the `interruptedException`.
is there anything else that would need to be set?

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:489]: preserving the root cause is the main thing.

File: src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java [hunk 8]
```diff
@@ -1132,4 +1133,10 @@ public static String encode(String s) {
         }
         return sb.toString();
     }
+
+    public static InterruptedIOException toInterruptedIOException(InterruptedException ex){
+        InterruptedIOException interruptedIOException = new InterruptedIOException();
+        interruptedIOException.initCause(ex);
+        return interruptedIOException;
+    }
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java:1141]: The copyright years need updating.

**Review by dfuch — APPROVED** [review]

**Comment by dfuch** [comment]
Actually, there's one more thing that we probably should do. The API documentation of `HttpResponse.BodySubscribers::ofInputStream` has an `@apiNote` that emphasizes the need to read the stream to exhaustion, or call close, in order to release all resource.
We probably should also add an `@apiNote` to document that the default implementation returned by that method throw `InterruptedIOException` if the thread is interrupted while reading from the stream.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 9]
```diff
@@ -481,7 +482,13 @@ private ByteBuffer current() throws IOException {
...
                     currentBuffer = currentListItr.next();
                 } catch (InterruptedException ex) {
-                    // continue
+                    try {
+                        close();
+                    } catch (IOException ignored) {
+                    }
+                    // Throw InterruptedIOException where the initCause is
+                    // set to the caught InterruptedException
+                    throw Utils.toInterruptedIOException(ex);
```
Comment (vyommani) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:491]: Don't you think you need to  restores the interrupt status of the thread by invoking the Thread.currentThread().interrupt() method and allowing the caller to detect the interrupt if it needs to ?

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:491]: By convention, any method that exits by throwing an InterruptedException clears interrupt status when it does so. Since we're throwing InterruptedIOException here, I don't think we need to set the interrupt flag. But the caller might do so if they want.

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:491]: Hmmm... Maybe a better solution would be to throw "ClosedByInterruptException" since we're also closing the input stream. And the specification of  "ClosedByInterruptException"  clearly state that the interrupt status of the thread will be preserved. Good question @vyommani .

Comment (vyommani) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:491]: I am not 100% sure but for me the best option is call close() and  just restore the interrupt status let the caller decide how to handle the interrupt.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 10]
```diff
@@ -481,7 +482,11 @@ private ByteBuffer current() throws IOException {
                     if (debug.on()) debug.log("Next Buffer");
                     currentBuffer = currentListItr.next();
                 } catch (InterruptedException ex) {
-                    // continue
+                    try {
+                        close();
+                    } catch (IOException ignored) {
+                    }
```
Comment (AlanBateman) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:488]: This will swallow the interrupt status, you' need to do Thread.currentThread().interrupt() before throwing the IOException.

**Comment by DarraghClarke** [comment]
Thanks for pointing out those issues, with the last few commits I've hopefully addressed everything, but let me know if there is anything else that need to be changed

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+                        .build();
+
+                // Send a httpRequest and assert the first response is received as expected
+                HttpResponse<InputStream> response = client.send(request, HttpResponse.BodyHandlers.ofInputStream());
+                String firstOutput = new String(response.body().readNBytes(FIRST_MESSAGE.getBytes().length));
+                assertEquals(firstOutput, FIRST_MESSAGE);
+
+                // countdown on latch, and assert that an IOException is throw due to the interrupt
+                interruptReadyLatch.countDown();
+                assertThrows(IOException.class, () -> response.body().readAllBytes(), "expected IOException");
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:166]: I suggest checking that the exception cause is the expected InterruptedException. If I'm not mistaken something like that should do it:

```
var thrown = assertThrows(IOException.class, () -> response.body().readAllBytes(), "expected IOException");
var cause = thrown.getCause();
assertTrue(cause instanceof InterruptedException, cause + " is not an InterruptedException");

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            } catch (Exception e) {
+                throw new RuntimeException(e);
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:168]: You might want to catch Throwable, then print the stack trace and call Asserts.fail here - because unhandled exceptions in threads might just be swallowed.

**Review by dfuch — CHANGES_REQUESTED** [review]

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 13]
```diff
@@ -162,10 +164,14 @@ static Thread createClientThread(CountDownLatch interruptReadyLatch, int port) {
...
 
                 // countdown on latch, and assert that an IOException is throw due to the interrupt
+                // and assert that the cause is a InterruptedException
                 interruptReadyLatch.countDown();
-                assertThrows(IOException.class, () -> response.body().readAllBytes(), "expected IOException");
-            } catch (Exception e) {
-                throw new RuntimeException(e);
+                var thrown = assertThrows(IOException.class, () -> response.body().readAllBytes(), "expected IOException");
+                var cause = thrown.getCause();
+                assertTrue(cause instanceof InterruptedException, cause + " is not an InterruptedException");
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:171]: Ah - and of course the test should verify that the thread interrupted status is set too, since we specified that :-)

```
var thread = Thread.currentThread();
assertTrue(thread.isInterrupted(), "Thread " + thread + " is not interrupted");
```

**Review by dfuch — APPROVED** [review]
LGTM - please make sure the test (and tier1 tier2) all pass before integrating

**Review by vyommani — COMMENTED** [review]
Looks OK to me.

**Review by vyommani — APPROVED** [review]

File: src/java.net.http/share/classes/java/net/http/HttpResponse.java [hunk 14]
```diff
@@ -1178,6 +1178,12 @@ public static BodySubscriber<Path> ofFile(Path file) {
          * the underlying HTTP connection to be closed and prevent it
          * from being reused for subsequent operations.
          *
+         * @implNote The read method of the default implementation returned
```
Comment (jaikiran) [inline src/java.net.http/share/classes/java/net/http/HttpResponse.java:1181]: Hello Darragh, should we instead reword this as:

> @implNote The {@code read} method of the {@code InputStream} returned by the default implementation of this method will throw an {@code IOException} with the {@link Thread#isInterrupted() thread interrupt status set} if the thread is interrupted while blocking on read. In that case, the request will also be cancelled and the {@code InputStream} will be closed.

Additionally, irrespective of whether or not we decide to reword this, this file will need an update for the copyright year.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 15]
```diff
@@ -481,7 +482,12 @@ private ByteBuffer current() throws IOException {
...
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:488]: I don't have specific opinion here and this is more of a question - should we be doing:

```
} catch (InterruptedException ex) {
    final IOException toThrow = new IOException(ex);
    try {
        close();
    } catch (IOException ignored) {
        toThrow.addSuppressed(ignored);
    }
    Thread.currentThread().interrupt();
    throw toThrow;    
}
```
i.e. should we be adding any failure to close() as a suppressed exception to the IOException that we throw?

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        server.start();
+    }
+
+    @AfterAll
+    void after() throws Exception {
+        server.stop(0);
+    }
+
+    @Test
+    public void test() {
```
Comment (jaikiran) [inline test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:83]: To keep this method relatively simpler, you could remove the top level try/catch block (which just catches and rethrows) and change this method's signature to `throws Exception`. If you decide to leave it in the current form, that's fine too.

File: test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java (hunk 6 above)
Comment (jaikiran) [reply test/jdk/java/net/httpclient/HttpResponseInputStreamInterruptTest.java:112]: I believe this change was missed and is pending?

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 9 above)
Comment (dmlloyd) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: InterruptedIOException is not equivalent to InterrupedException in this way AFAIK. The thread interrupt status should be reset if _any_ exception other than InterruptedException is thrown AFAIK, otherwise thread interruption is likely to be lost.

File: src/java.net.http/share/classes/java/net/http/HttpResponse.java (hunk 14 above)
Comment (dfuch) [reply src/java.net.http/share/classes/java/net/http/HttpResponse.java:1181]: Agreed but it should be `{@linkplain Thread#isInterrupted() thread interrupt status set}`

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 15 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:488]: I am not sure whether that would be helpful or confusing.

Comment (DarraghClarke) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:488]: I think that's a good point, I don't really have a strong opinion one way or the other on it but would be curious if anyone else does?

**Review by dfuch — APPROVED** [review]

**Review by jaikiran — APPROVED** [review]
Thank you for these changes Darragh. Looks fine to me. There was one review comment by Daniel where he noted that the fields of the `Handler` in the test should be `final` (which I agree with). From what I can see, that hasn't been addressed in the updates. In the current test, it doesn't functionally impact anything, so it is OK if you want to leave it in this form.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 15 above)
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:487]: Let's leave it in the current form then.
